core/dnsserver/watch.go

```diff
@@ -0,0 +1,166 @@
+package dnsserver
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"sync"
+
+	"github.com/miekg/dns"
+
+	"github.com/coredns/coredns/pb"
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/watch"
+)
+
+// watcher contains all the data needed to manage watches
+type watcher struct {
+	changes watch.Chan
+	counter int64
+	watches map[string]watchlist
+	plugins []watch.Watchable
+	mutex   sync.Mutex
+}
+
+type watchlist map[int64]pb.DnsService_WatchServer
+
+func newWatcher(zones map[string]*Config) *watcher {
```

Comment:
I originally tried to do it that way. We can probably move code to a pkg or something. Honestly it's been a while since I did that and I don't recall the reasons exactly, other than that it needs to be in the gRPC service and so at its core in the grpc server.